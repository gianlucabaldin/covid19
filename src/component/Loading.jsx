import React from 'react';
import ClipLoader from 'react-spinners/ClipLoader';

const Loading = ({ loading = true }) => (
  <ClipLoader
    color="#303f9f"
    loading={loading}
    size={40}
    data-id="loading-spinner"
  />
);

export default Loading;
