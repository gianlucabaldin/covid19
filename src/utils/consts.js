export const DEFAUL_MAX_DATA_SIZE = 30;

export const API_ITALY_HYSTORICAL =
  'https://api.covid19api.com/total/dayone/country/italy';

export const API_ITALY_HYSTORICAL_APIFY =
  'https://api.apify.com/v2/datasets/CUdKmb25Z3HjkoDiN/items?format=json&clean=1';
export const API_ITALY_HYSTORICAL_APIFY_SHORT_URL = 'api.apify.com';

export const API_ITALY_REGION =
  'https://openpuglia.org/api/?q=getdatapccovid-19';

export const OPEN_PUGLIA_API = 'openpuglia.org';
export const COVID_19_API = 'covid19api.com';

export const STATUS = {
  CONFIRMED: 'CONFIRMED',
  RECOVERED: 'RECOVERED',
  DEATHS: 'DEATHS',
  ACTIVES: 'ACTIVES',
  INTENSIVE_THERAPY: 'INTENSIVE_THERAPY',
  TOTAL_HOSPITALIZED: 'TOTAL_HOSPITALIZED',
  TOTAL_POSITIVE: 'TOTAL_POSITIVE',
  TAMPONI: 'TAMPONI',
  TOTAL_CASES: 'TOTAL_CASES',
  DECEASED: 'DECEASED',
  NEW_DAILY_POSITIVE: 'NEW_DAILY_POSITIVE',
  DAILY_DECEASED: 'DAILY_DECEASED',
  DAILY_SWABS: 'DAILY_SWABS',
};
