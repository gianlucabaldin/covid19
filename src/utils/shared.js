/* eslint-disable import/prefer-default-export */

/**
 * Check if the value is a number and more than a thousand, in that case
 * call the toLocalString() witch inserts (eventually) the separator, used
 * in Summary
 * I.E. for italian it inserts the "." ad thousand separator
 * @param {number || string} val the value to insert the separator to if > thousand
 */
export const getLocalizedValue = (val) => {
  if (typeof val === 'number') {
    return Number(val).toLocaleString();
  }
  return val;
};
