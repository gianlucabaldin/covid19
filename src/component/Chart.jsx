import React, { useState } from 'react';
import moment from 'moment';
import {
  Hint,
  VerticalGridLines,
  HorizontalGridLines,
  LineMarkSeries,
  DiscreteColorLegend,
  XAxis,
  YAxis,
} from 'react-vis/dist';
import { FlexibleWidthXYPlot } from 'react-vis/dist/make-vis-flexible';
import { useTranslation } from 'react-i18next';
import { makeStyles } from '@material-ui/core';
import Loading from './Loading';

const useStyles = makeStyles((theme) => ({
  legend: {
    marginLeft: '40%',
    padding: 0,
    [theme.breakpoints.down('sm')]: {
      marginLeft: 'inherit',
      textAlign: 'center',
    },
  },
}));

const Chart = ({ series, status, loading = true }) => {
  const classes = useStyles();

  const initialHintValue = {
    data: {},
    over: false,
    status: '',
  };
  const [hint, setHint] = useState(initialHintValue);

  const { t } = useTranslation();

  const mouseOver = (datapoint, datapointStatus) => {
    setHint({ data: datapoint, over: true, status: datapointStatus });
  };

  const mouseOut = (datapoint) => {
    setHint({ data: datapoint, over: false, status: '' });
  };

  const getHintSection = () => {
    return hint.over ? (
      <Hint value={hint.data}>
        <div
          style={{
            background: 'white',
            padding: 5,
            border: '1px solid black',
            borderRadius: 10,
            color: '#303f9f',
          }}
        >
          <span style={{ fontWeight: 'bold', textTransform: 'capitalize' }}>
            {t('chart.day')}:{' '}
          </span>{' '}
          {moment(hint.data.x).format('DD/MM')} <br />
          <span style={{ fontWeight: 'bold', textTransform: 'capitalize' }}>
            {t(hint.status.toString().toLowerCase()).replace('_', ' ')}:
          </span>{' '}
          {hint.data.y.toLocaleString()}
        </div>
      </Hint>
    ) : null;
  };

  const getLegend = () => {
    const array = [];
    if (status && status.length > 0) {
      status.forEach((el) => {
        array.push(
          <span style={{ fontWeight: 'bold', textTransform: 'capitalize' }}>
            {t(el.toString().toLowerCase()).replace('_', ' ').toUpperCase()}
          </span>,
        );
      });
    }
    return array;
  };
  return loading ? (
    <div>
      <Loading />
    </div>
  ) : (
    <div data-test-id="chart">
      <DiscreteColorLegend
        strokeWidth={2}
        orientation="vertical"
        items={getLegend()}
        // style={{ marginLeft: '40%', padding: 0 }}
        className={classes.legend}
      />
      <FlexibleWidthXYPlot height={250} xType="time">
        <HorizontalGridLines />
        <VerticalGridLines />

        {series &&
          series.length > 0 &&
          series.map((serie, i) => (
            <LineMarkSeries
              getNull={(d) => d.y !== undefined} // data hole / attribute changes
              // curve="curveMonotoneX"
              data={serie || []}
              onValueMouseOver={(val) => mouseOver(val, status[i])}
              onValueMouseOut={mouseOut}
              style={{
                strokeWidth: '3px',
              }}
              lineStyle={{ stroke: 'red' }}
              markStyle={{ stroke: 'blue' }}
              size={3} // mark size, default=5
              fill="red" // the color inside the mark
            />
          ))}
        <XAxis
          title="day"
          tickFormat={(value) => moment(value).format('DD/M')}
          tickTotal={10}
        />
        <YAxis title="number" position="end" />
        {/* the hint popup shown when user point a mark with the mouse */}
        {getHintSection(hint.over)}
      </FlexibleWidthXYPlot>
    </div>
  );
};

export default Chart;
