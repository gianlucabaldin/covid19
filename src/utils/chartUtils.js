// const DEFAUL_MAX_DATA_SIZE = 20;

import { DEFAUL_MAX_DATA_SIZE } from './consts';

// In order to have a "readable" chart, reduce if too much input values
export const reduceData = (data, maxSize = DEFAUL_MAX_DATA_SIZE) => {
  const { length } = data;
  if (length > maxSize) {
    const y = Math.round(length / maxSize);
    data = data.filter((el, i) => Math.round(i % y) === 0);
  }
  return data;
};
