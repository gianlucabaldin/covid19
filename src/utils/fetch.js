import { mockItalyHistoricalResponseApify } from './mock/mockItalyHistoricalResponse_Apify';
import { mockItalyRegionsResponse } from './mock/mockItalyRegionsResponse';
import { API_ITALY_HYSTORICAL_APIFY, API_ITALY_REGION } from './consts';

export const fetchItalyHistoricalAllApify = async (mock = false) => {
  if (mock) {
    console.log('mock = true');
    return mockItalyHistoricalResponseApify;
  }
  console.log('mock = false');
  return fetch(API_ITALY_HYSTORICAL_APIFY).then((res) => res.json());
};

export const fetchItalyRegion = async (mock = false) => {
  if (mock) {
    return mockItalyRegionsResponse;
  }
  return fetch(API_ITALY_REGION).then((res) => res.json());
};
