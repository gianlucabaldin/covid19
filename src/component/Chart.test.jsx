import React from 'react';
import { shallow } from 'enzyme';
import { render } from '@testing-library/react';
import Chart from './Chart';
import Loading from './Loading';
import { STATUS } from '../utils/consts';

describe('Chart', () => {
  let wrapper = shallow(<Chart />);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('with running spinner', () => {
    expect(wrapper.contains(<Loading />)).toEqual(true);
  });

  describe('with spinner', () => {
    it('is not running', () => {
      wrapper = shallow(<Chart loading={false} />);
      expect(wrapper.contains(<Loading />)).toEqual(false);
    });

    it('shows the legend', () => {
      const { getByText } = render(
        <Chart loading={false} status={[STATUS.DAILY_SWABS]} />,
      );
      expect(getByText('DAILY SWABS')).toBeInTheDocument();
    });
  });
});
