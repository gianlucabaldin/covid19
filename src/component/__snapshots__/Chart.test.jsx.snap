// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart renders correctly 1`] = `
<div
  style={
    Object {
      "width": "100%",
    }
  }
>
  <Loading />
</div>
`;
